Source/Factories/JSONFactory.swift

```diff
@@ -0,0 +1,105 @@
+    case missing(JSONObject, String)
+}
+
+/// A factory that wraps `Component` builder closures (ViewFactoryBuilder, 
+/// `WrapperFactoryBuilder` & `ClusterFactoryBuilder`) and uses them to produce `Component`s
+public final class JSONFactory {
+    
+    private enum Key {
+        static let type = "type"
+        static let meta = "meta"
+        static let children = "children"
+    }
+    
+    /// A factory closure to build a view `Component`
+    public typealias ViewFactoryBuilder = (ComponentMeta?) -> Component
+    
+    /// A factory closure to build a wrapper `Component`
+    public typealias WrapperFactoryBuilder = (Component, ComponentMeta?) -> Component
+    
+    /// A factory closure to build a cluster `Component`
+    public typealias ClusterFactoryBuilder = ([Component], ComponentMeta?) -> Component
+    
+    private var viewFactory: [String: ViewFactoryBuilder] = [:]
+    private var wrapperFactory: [String: WrapperFactoryBuilder] = [:]
+    private var clusterFactory: [String: ClusterFactoryBuilder] = [:]
+    
+    /// Initialize a new JSONFactory
+    public init() {
+        // Empty but needed to be initialized from other modules
+    }
+    
+    /// Registers a new `ViewFactoryBuilder` which will be used when producing the component
+    ///
+    /// - Parameters:
+    ///   - type: a string identifying this factory type
+    ///   - factoryBuilder: a `ViewFactoryBuilder` to build a `Component`
+    public func register(with type: String,
+                         factoryBuilder: @escaping ViewFactoryBuilder) {
+        viewFactory[type] = factoryBuilder
+    }
+    
+    /// Registers a new `WrapperFactoryBuilder` which will be used when producing the component
+    ///
+    /// - Parameters:
+    ///   - type: a string identifying this factory type
+    ///   - factoryBuilder: a `WrapperFactoryBuilder` to build a `Component`
+    public func register(with type: String,
+                         factoryBuilder: @escaping WrapperFactoryBuilder) {
+        wrapperFactory[type] = factoryBuilder
+    }
+    
+    /// Registers a new `ClusterFactoryBuilder` which will be used when producing the component
+    ///
+    /// - Parameters:
+    ///   - type: a string identifying this factory type
+    ///   - factoryBuilder: a `ClusterFactoryBuilder` to build a `Component`
+    public func register(with type: String,
+                         factoryBuilder: @escaping ClusterFactoryBuilder) {
+        clusterFactory[type] = factoryBuilder
+    }
+    
+    /// Produces a `Component` from a given `JSONObject`, which has one mandatory key: `type`
+    ///
+    /// - Parameter json: the `JSONObject` to be used
+    /// - Returns: An optional `Component`
+    /// - Throws: `JSONFactoryError` when a mandatory key is missing
+    public func component(from json: JSONObject) throws -> Component? {
+        guard let type = json[Key.type] as? String else {
+            throw JSONFactoryError.missing(json, Key.type)
```

Comment:
I mentioned this on last meeting and they said we don't include fallback components for now? We could implement it here though.